crates/runtime/src/datafusion/cluster/mod.rs

```diff
@@ -0,0 +1,276 @@
+use ballista_executor::executor::Executor;
+use ballista_executor::metrics::LoggingMetricsCollector;
+use ballista_scheduler::cluster::BallistaCluster;
+use ballista_scheduler::config::SchedulerConfig;
+use ballista_scheduler::scheduler_process;
+use ballista_scheduler::scheduler_server::SchedulerServer;
+use datafusion::execution::SessionStateBuilder;
+use datafusion::prelude::SessionConfig;
+use datafusion_proto::protobuf::{LogicalPlanNode, PhysicalPlanNode};
+use futures::TryFutureExt;
+use runtime_object_store::registry::default_runtime_env;
+use snafu::ResultExt;
+use std::env;
+use std::sync::Arc;
+use tokio::net::TcpListener;
+use uuid::Uuid;
+
+pub mod codec;
+pub mod common;
+pub mod config;
+pub mod physical_plan;
+
+/// Creates & binds a Ballista scheduler to the Runtime handle, then updates status
+pub async fn initialize_cluster_scheduler(rt: &Arc<Runtime>) -> crate::Result<()> {
+    let scheduler = create_scheduler_server(rt).await?;
+
+    rt.df
+        .bind_scheduler_server(Arc::new(scheduler))
+        .map_err(|e| FailedToStartClusterScheduler {
+            source: Box::new(e),
+        })?;
+
+    rt.status
+        .update_cluster("scheduler", ComponentStatus::Ready);
+
+    Ok(())
+}
+
+/// Creates a Ballista executor, binds it to the `Runtime` handle, and returns its configured
+/// work loop as a future
+pub async fn initialize_cluster_executor(
+    rt: Arc<Runtime>,
+) -> crate::Result<impl Future<Output = crate::Result<()>>> {
+    executor_bind_app(&rt, rt.config.cluster.scheduler_url.to_string()).await?;
+
+    let runtime_handle = Arc::clone(&rt);
+
+    let runtime_producer: RuntimeProducer =
+        Arc::new(move |_cfg| Ok(Arc::clone(&runtime_handle.df.ctx.runtime_env())));
+
+    let config_producer: ConfigProducer = Arc::new(move || {
+        SessionConfig::new_with_ballista().with_option_extension(SpiceClusterConfig::default())
+    });
+
+    let work_dir = rt
+        .df
+        .temp_directory
+        .clone()
+        .unwrap_or(env::temp_dir().to_string_lossy().to_string());
+
+    let scheduler_connection =
+        create_grpc_client_connection(rt.config.cluster.scheduler_url.clone().to_string())
+            .await
+            .map_err(|_| FailedToStartClusterExecutor {
+                source: format!(
+                    "Unable to connect to scheduler at {}",
+                    rt.config.cluster.scheduler_url
+                )
+                .into(),
+            })?;
+
+    let scheduler = SchedulerGrpcClient::new(scheduler_connection)
+        .max_encoding_message_size(usize::MAX)
+        .max_decoding_message_size(usize::MAX);
+
+    // Try to bind the same flight port Spice usually does, but if we cannot, bind a different
+    // port to allow for easy local deployments
+    let default_grpc_binding = TcpListener::bind(rt.config.flight_bind_address)
+        .await
+        .and_then(|l| l.local_addr());
+
+    let dynamic_grpc_binding = TcpListener::bind("0.0.0.0:0")
```

Comment:
(can come in a follow up): we shouldn't attempt to bind to the random address unless the first one fails, and then we should probably only bind to localhost.